File: 529. Minesweeper.java
class Solution {//Time and space of O(MN)
    int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0},{1,1},{1,-1},{-1,1},{-1,-1}};
    public char[][] updateBoard(char[][] board, int[] click) {
        
        int row = click[0];
        int col = click[1];
        //base 
        if(board[row][col] == 'M'){
            board[row][col] = 'X';
            return board;
        }
        dfs(board,row,col);
        
        return board;
    }
    public void dfs(char[][] board, int i, int j){
        //Check if the i and j are valid
        if(i<0 || i>=board.length || j< 0 || j>= board[0].length || board[i][j] != 'E'){
            return;
        }
        int mines = getMines(board ,i,j);
        
        if(mines == 0){
            board[i][j] = 'B';
            for(int[] dir:dirs){
              int newRow = i + dir[0];
              int newCol = j + dir[1];
              dfs(board,newRow, newCol);
         }
        }else{
            board[i][j] = (char)(mines+'0') ;
        }  
    }
    public int getMines(char[][] board ,int i, int j){
        int count = 0;
        for(int[] dir:dirs){
              int newRow = i + dir[0];
              int newCol = j + dir[1];
              if(newRow<0 || newRow>=board.length || newCol< 0 || newCol>= board[0].length){
                  continue;
              }
            if(board[newRow][newCol] == 'M') count++;
         }
        return count;
    }
}

File: 529-Minesweeper.cpp
/* 
    Time Complexity                              :  O(N)
    Space Complexity                             :  O(1)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  

class Solution {
private: 
    vector<vector<int>> dirs{{-1,-1},{-1,0},{-1,1},{0,1},{1,1},{1,0},{1,-1},{0,-1}};
public:
    vector<vector<char>> updateBoard(vector<vector<char>>& board, vector<int>& click) {
        int n = board.size(), m = board[0].size();
        queue<vector<int>> q;
        q.push(click);
        while(!q.empty()) {
            vector<int> f = q.front();q.pop();
            int cx = f[0], cy = f[1];
            if(board[cx][cy] == 'B') {
                continue;
            } else if(board[cx][cy] == 'M') {
                board[cx][cy] = 'X';
                return board;
            }
         
            
            int mineCount = 0;
            for(auto dir : dirs) {
                int nx = cx + dir[0];
                int ny = cy + dir[1];
                
                if(nx >= 0 and nx < n and ny >= 0 and ny < m and board[nx][ny] == 'M') {
                    mineCount++;
                }
            }
            
            if(board[cx][cy] == 'E' and mineCount == 0) {
                board[cx][cy] = 'B';
                for(auto dir : dirs) {
                    int nx = cx + dir[0];
                    int ny = cy + dir[1];
                
                    if(nx >= 0 and nx < n and ny >= 0 and ny < m) {
                        q.push({nx,ny});
                    }
                }
            } else {
                board[cx][cy] = '0' + mineCount;
            }
        }
        
        return board;
    }
};